### PR TITLE
interfaces: add `snap-interfaces-requests-control` interface

### PR DESCRIPTION
**Pull request by olivercalder**
Add the `snap-prompting-control` interface to allow prompting clients to be packaged as snaps. This interface grants access to the prompting-related notice types `interfaces-requests-prompt` and `interfaces-requests-rule-update`, as well as to the `/v2/system-info` endpoint, which lets clients identify whether prompting is supported and enabled. In the future, this interface will also grant access to the prompting API endpoints.

The corresponding review-tools PR can be found here: https://code.launchpad.net/~alexmurray/review-tools/+git/review-tools/+merge/470317

This work is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-17798.

### CODE REVIEW

**Comment by olivercalder** [comment]
This PR needs security review, CC: @alexmurray.

In order to publish the `prompting-client` snap being worked on by @sminez, that snap will need approval to use and autoconnect this interface. Please let me know if anything about this PR needs to be adjusted to allow this to happen, or if a PR should be opened elsewhere to make the store aware of this interface.

Note as well the previous exception for `snapd-desktop-integration`. My impression is that going forward we want to go through the proper channels for snaps which want to use this interface, so I removed that exception, and don't think we want a new one for `prompting-client`.

**Comment by olivercalder** [comment]
CC @robert-ancell as well

File: daemon/api_notices.go
Comment (zyga) [inline daemon/api_notices.go:47]: Curious, why is write access more open than read access?

File: daemon/api_notices.go
Comment (bboozzoo) [reply daemon/api_notices.go:46]: https://github.com/canonical/snapd/pull/14272/files#diff-5a0f5ef8cf432061d3b8999f592aa2b22c109b1b2aeda9e7cce4916ba5188f2dR104 checks whether the snap plugs appropriate interface to be able to read notices of given type(s)

File: daemon/api_notices.go
Comment (bboozzoo) [reply daemon/api_notices.go:47]: the names aren't great perhaps, openAcces{} is open for all requests but on the system socket only, while interfaceOpenAccess{} allows snap-snap socket and checks the plugged interfaces

File: overlord/state/notices_test.go
Comment (bboozzoo) [inline overlord/state/notices_test.go:334]: why not have all 3 types used here?

**Review by bboozzoo — APPROVED** [review]
LGTM

File: daemon/api_notices.go
Comment (olivercalder) [inline daemon/api_notices.go:38]: See here @zyga, this is what specifies which interface grants access to which notice types.

File: overlord/state/notices_test.go
Comment (olivercalder) [reply overlord/state/notices_test.go:334]: The idea was to just check some of the types. But no particular reason not to include `RefreshInhibitNotice` as well. This change is from https://github.com/canonical/snapd/pull/14269 by the way, it will be eaten when that PR merges and this one is rebased.

File: overlord/state/notices.go
Comment (alexmurray) [inline overlord/state/notices.go:211]: To follow the existing convention here, should this instead be "requests-prompt-notice"? And similarly for the one below?

**Review by alexmurray — APPROVED** [review]
LGTM - just one minor nitpick on a naming convention.

File: overlord/state/notices.go
Comment (olivercalder) [reply overlord/state/notices.go:211]: From these, I think the convention is not to have "-prompt" at the end of the notice type string:https://github.com/canonical/snapd/pull/14272/files#diff-4fc1b33549a08f2c1e8c5ae6ebdd8a0b859cc5e93b7208bc6656e02c208a9183R193-R211

It's a good question whether these should simply be "request-prompt" and "request-rule" instead of "interfaces-requests-prompt" and "interfaces-requests-rule-update". These changes are actually from #14269 though, so I'd rather address them there :)

File: tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml
Comment (ZeyadYasser) [inline tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml:224]: Maybe a simple standalone spread test would give better coverage? something like `tests/main/interfaces-snap-refresh-observe` .

Note, the spread test checks were updated that if a test named `interfaces-snap-prompting-control` exists it wouldn't complain about it being absent from `test-snapd-policy-app-consumer`.

This can be done in a follow up PR, it is not urgent.

**Review by ZeyadYasser — APPROVED** [review]
Great work, Thanks!

File: overlord/state/notices.go
Comment (sminez) [reply overlord/state/notices.go:211]: would that change be backwards compatible / would we be able to pre-emptively request the newer name as well without anything breaking? Currently we have to specify the existing `interfaces-requests-prompts` notice type in our request to long poll for notices.

Comment (olivercalder) [reply overlord/state/notices.go:211]: We'll leave the `interfaces-requests-prompt` and `interfaces-requests-rule-update` names unchanged.

For the record though, clients should be able to poll for both old and new names, if we ever do want to change something:

```
/v2/notices?types=interfaces-requests-prompt&types=interfaces-requests-prompt-update
```

For example. Snapd will ignore the unrecognized names and return any notices for the types which are recognized (and allowed by the connected interfaces).

File: tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml
Comment (olivercalder) [reply tests/lib/snaps/test-snapd-policy-app-consumer/meta/snap.yaml:224]: Great point, thank you! I'll add this now, since we're blocked by GH actions at the moment anyway

File: daemon/api_notices_test.go
Comment (pedronis) [inline daemon/api_notices_test.go:378]: these changes are a bit strange, what I think would make sense it to parametrize the test and run both the old variant and again with the new interface

File: interfaces/builtin/snap_prompting_control.go
Comment (pedronis) [inline interfaces/builtin/snap_prompting_control.go:40]: wouldn't snap-interfaces-requests-control be a more consistent name now?

**Review by pedronis — COMMENTED** [review]
thx, couple of comments

**Review by ZeyadYasser — APPROVED** [review]

**Comment by olivercalder** [comment]
We have a new plan after discussion with Desktop/Samuele on Friday regarding a way to ensure that prompting client(s) (which are user daemon snaps) are started when the prompting flag is enabled.

At the moment, when the prompting client starts up, it checks snapd to see if it supports prompting, and if it does not, it shuts down. There is not an existing way for the client to restart itself when the prompting flag changes (or when snapd restarts). The only solution at the moment would be for the client to always run and repeatedly poll `/v2/system-info` and/or the prompting endpoints. In addition to being a drag on the performance of snapd, this means the prompting client would be running even when prompting is not enabled, and the user might kill or disable the service, either because they don't know what it is or don't want it running all the time for no reason (assuming they have prompting disabled). If prompting is later enabled, there is no mechanism to ensure that the prompting client service is restarted/re-enabled.

So, the plan: add a `"handler-service"` attribute to the `snap-prompting-control` interface which maps to a list of service names within the snap which should be "ensured to be running" when snapd starts up with prompting supported/enabled. I say "ensured to be running" since there are a few complexities:
- Prompting may have been enabled/running when snapd restarts, and the services may have been killed or disabled, so we want the resulting service status to be idempotent across snapd restarts
- We don't want to blanket restart the systemd services associated with the apps, since that would presumably kill in-flight prompts which the user is interacting with when snapd restarts
- Services may have been explicitly disabled, so we should also ensure that the services are enabled (in addition to running) when prompting is supported/enabled, so that they automatically restart (?) if they crash
- We may want to disable the services when snapd starts and prompting is not supported/enabled
- I don't think this should be a required attribute, since some snaps may wish to be a rules management client (like the Security Center) without being a prompt handler client

I'd like to take the following steps:
1. Add the `"handler-service"` attribute in this PR, along with checks that the listed service(s) match the names of daemon apps within the snap
2. Move forward with reviewing/merging this PR so that the corresponding review-tools PR can be reviewed and snaps can be published with this interface sooner rather than later
3. In a follow-up PR, add the internal implementation of what snapd does with the services listed under the `"handler-service"` attribute, since there is some design work to be done here

**Comment by sminez** [comment]
@olivercalder from the Desktop side those steps sound good to me. We should book in some time to talk about the expected / required behaviour of the client around restarts and what changes you want from us to help this run smoothly. The way the client runs as two binaries: the daemon polling snapd and replying to prompts and the Flutter UI, so in theory we would be able to have the UI remain open across a restart of the daemon.

**Comment by pedronis** [comment]
> So, the plan: add a "handler-service" attribute to the snap-prompting-control interface which maps to a list of service names

my proposal was for this to list *one* service.  one service which should be the main service dedicated to reply to prompts, notice that once that service is running if we have other mechanisms for it to manage other services in the snap

**Comment by olivercalder** [comment]
Had good discussion with @zyga and @bboozzoo about the interface and proposed attribute and their relation to prompt clients. Here are some of the key takeaways:
- The presence of an interface connection for the snap-prompting-control interface cannot be included in the system key along with checks for the AppArmor kernel and parser support, since connections aren't known at that point during snapd startup
- Furthermore, we don't want there to be any difference between checks for the presence of a client at startup and checks for the presence of a client later during runtime, when we need to adjust behavior due to a prompting client being installed/uninstalled
- So, when snapd starts up and sees that the prompting flag is enabled and the AppArmor kernel and parser features support prompting, then prompt prefixes should be included in the relevant AppArmor profiles (as is the case now) regardless of the presence of a prompt handler service
- If a snap is installed with a `snap-prompting-control` plug connected which includes a valid `"handler-service"` attribute, then snapd should default to denying requests after a timeout (around 1 minute?) unless they are replied to by the client
- If no such connected plug with the `"handler-service"` attribute, then snapd should consult the existing rules, and if the request is matched by an existing rule, it should be handled accordingly, otherwise, it should be allowed by default (since there is no client present to reply)
  - This also makes prompting viable on servers where there would be no interactive prompt client: a user could create allow and deny rules precisely as they desire, then let prompting run, doing as they declared when relevant, otherwise behaving as if prompting were not running
- Snapd should not be responsible for starting clients given by the `"handler-service"` attribute, this should be the responsibility of the client itself
  - If a user manually disabled a prompt client service, we should not silently re-enable it
  - If a user (or malicious snap) killed a prompt client service, it should be restarted automatically by systemd, and even if not, the interface is still connected with the `"handler-service"` attribute, so snapd will default to deny if the client is inoperable
- If snapd is in fact responsible for starting prompt client services, this needs to be done during runtime, with the most obvious time being when a request is received from the kernel
  - The snapd user session (I don't know the correct term for this) starts up when a user logs in, which may be after snapd has already started, so we really can't tell the user session anything until the last second, in case it hasn't been started yet

Some notes from discussion of user daemons with @Meulengracht:
- User daemons is experimental, but should all be working fine aside from tracking of enablement status across refreshes, for which a PR is outstanding
  - Once landed, all user daemons functionality should be working
- If a user service doesn't need to be running immediately after it is installed, it can be marked as such, and only enabled when needed
- When snapd restarts, it shouldn't affect running services
- A client using a user daemon requires the experimental `UserDaemons` feature to be enabled
  - Will all snap prompt clients use user daemon services as prompt handlers?
    - Probably can't assume this
  - Should enablement of `AppArmorPrompting` experimental feature trigger enablement of `UserDaemon` experimental feature as well?
    - Probably not, but then it's the client's responsibility to enable user daemons when installing the client
- Once the PR for tracking service enablement across snapd restarts is merged, user daemons feature can probably be enabled by default
- Can only have one user service per snap per user

**Comment by olivercalder** [comment]
CC @ernestl

**Comment by olivercalder** [comment]
New tentative plan after discussion with Oliver Smith, @pedronis, @sminez, @zyga, @ernestl: simplify as much as possible, no runtime checks, make the flag a toggle for "is prompting running", and only allow it to be turned on if we support prompting and there's a client ready to receive prompts.

Pieces:
- Make `"apparmor-prompting"` experimental feature flag only able to be turned on if prompting is supported
  - Do we want to also check that there's a snap with the `snap-prompting-control` interface connected and a `"handler-service"` attribute?
    - Would be simpler to not
      - If in the future we don't use a snap with `snap-prompting-control` interface as prompt handler, there's no lock-in, and prompting will be supported even if there's no prompt client snap
    - Would be "safer" from a UX perspective to check, but there is complexity around checking, and when starting up or during refresh, connections are not always present
- When `"apparmor-prompting"` flag is successfully enabled, ensure that all services annotated with `"handler-service"` attributes are enabled
  - Then corresponding user service be started when user logs in, and should start the service immediately
  - This should be done in `overlord/configstate/configcore/prompting.go` as soon as the above checks pass
- Consolidate prompting support and enabled values into one
  - This has already been done for the system key, should just be a matter of changing the enablement to succeed if prompting is supported
- Add timeout to each prompt if no client queries it
  - Default to short timeout of 10s
  - When client gets prompt details via `/v2/interfaces/requests/prompts/{id}` (or via `/v2/interfaces/requests/prompts`, really this timeout could all be handled internally to the `requestprompts` package, so whenever the prompt is accessed internally its timeout is increased), add long timeout (1 min?) to timeout for that prompt, as this shows there is a client acting on prompt details
    - No timeout checks on notices API, as this is complicated to check (notices API used for many things) and doesn't really show intent to action prompts as clearly as getting prompt details does
  - When a prompt times out, add a `logger.Notice` (along with a notice for the prompt stating that the prompt timed out), so the user has decent opportunities to figure out why their applications are hanging

File: daemon/api_notices_test.go
Comment (olivercalder) [reply daemon/api_notices_test.go:378]: Yes, will do!

**Comment by olivercalder** [comment]
Complexity remains regarding how to handle prompts for the `root` user. Unless someone is logged in as root in a graphical session with the client running (unclear exactly if/how the userd session runs for root), all prompts for the root user will remain unactioned forever, and timeout with a deny. This is obviously bad. But no non-root users have permission to see the prompt notices addressed to root, nor the prompt details, nor the authority to act on them even if they could see them.

The "easy" solution would be to auto-allow prompts originating from the root user. This actually seems rather reasonable to me, since if a user is manually running a snap as root, that user already has elevated privileges and could circumvent prompting anyway, so in effect, turning off prompting for root is not a real decrease in security (and not a change from the current status quo). Also, most snaps which are run as root are services started by systemd, and by definition these are headless services not associated with any particular user or session, so it makes sense that prompting would not work/apply to these. This would also solve one of the scary possibilities of enabling prompting on a server and thus making all services hang when they try to use a prompt-enabled interface. What do you think of this @zyga ?

**Comment by sminez** [comment]
> Add timeout to each prompt to auto-deny if no client queries it

This gets a little problematic in the case where a snap is multi-threaded and generates multiple prompts at once. With the current implementation we eagerly pull prompt details for each prompt that we receive a notice for but then present them to the user serially, so the first prompt would receive the full timeout but others would end up with a greatly reduced one.

> so whenever the prompt is accessed from an external method its timeout is increased

Is that what this is intended to address? I'm reading this as the client would need to poke prompts it was intending to action as a keep-alive mechanism. Is that correct?

**Comment by zyga** [comment]
@olivercalder I think this is sensible and buys us time to make it stricter later.

**Comment by olivercalder** [comment]
Fow now, I'll land the interface as-is, and add the attribute and client interactions in a follow-up PR.

**Comment by olivercalder** [comment]
Importantly, this interface has been renamed to `snap-interfaces-requests-control` from its original name of `snap-prompting-control`. The former matches the API endpoints, and the latter is rather vague. CC @sminez on this change, you'll need to adjust the interface name in the client snapcraft.yaml.

**Review by pedronis — APPROVED** [review]
thanks, this needs a quick review from somebody else from snapd

File: tests/main/interfaces-snap-interfaces-requests-control/task.yaml
Comment (zyga) [inline tests/main/interfaces-snap-interfaces-requests-control/task.yaml:17]: Perhaps setting locale to C.UTF8 would be more general?

File: tests/main/interfaces-snap-interfaces-requests-control/task.yaml
Comment (zyga) [inline tests/main/interfaces-snap-interfaces-requests-control/task.yaml:20]: Is edge relevant here?

Don't we have test-snapd-jq somewhere?

File: tests/main/interfaces-snap-interfaces-requests-control/task.yaml
Comment (zyga) [inline tests/main/interfaces-snap-interfaces-requests-control/task.yaml:62]: I presume all this is coming once the internals are ready.

**Review by zyga — APPROVED** [review]
I left a few comments inline. One possibly small tweak is to use non stock jq because I think we had a test-snapd- variant somewhere.

File: tests/main/interfaces-snap-interfaces-requests-control/task.yaml
Comment (olivercalder) [reply tests/main/interfaces-snap-interfaces-requests-control/task.yaml:62]: Yes, I'll wire this up once I rebase the chain in a bit. Fun times.

File: tests/main/interfaces-snap-interfaces-requests-control/api-client/bin/api-client.py
Comment (olivercalder) [reply tests/main/interfaces-snap-interfaces-requests-control/api-client/bin/api-client.py:1]: I'll give this a try, if it doesn't work though I'll leave it as-is.

File: tests/main/interfaces-snap-interfaces-requests-control/task.yaml
Comment (olivercalder) [reply tests/main/interfaces-snap-interfaces-requests-control/task.yaml:20]: Seems like we use a mix of standard `jq`, `--devmode`, and `--edge` elsewhere in snapd. I don't believe the `--edge` is relevant here, I'll remove it and test.

```
$ grep -r test-snapd-jq                 
tests/lib/snaps/store/test-snapd-jq-core24/snapcraft.yaml:name: test-snapd-jq-core24
tests/lib/prepare.sh:        snap install --devmode --edge test-snapd-jq-core24
tests/lib/prepare.sh:        snap alias test-snapd-jq-core24.jq jq
tests/lib/prepare.sh:    snap remove --purge test-snapd-jq-core24
```

We maybe could use `test-snapd-jq-core24`, but we use standard `jq` (from mvo) in most places, so I'd rather stick to it.

File: tests/main/interfaces-snap-interfaces-requests-control/task.yaml
Comment (olivercalder) [reply tests/main/interfaces-snap-interfaces-requests-control/task.yaml:17]: I don't see `C.UTF8` as an option here: https://docs.python.org/3/library/codecs.html#standard-encodings

Here's what I looked at:
1. https://docs.python.org/3/using/cmdline.html#envvar-PYTHONIOENCODING
2. https://docs.python.org/3/library/stdtypes.html#str.encode
3. https://docs.python.org/3/library/codecs.html#standard-encodings

Comment (olivercalder) [reply tests/main/interfaces-snap-interfaces-requests-control/task.yaml:17]: Default is `utf-8` though, so here we're forcing it so it doesn't try to be clever.

File: interfaces/builtin/snap_interfaces_requests_control.go
Comment (ZeyadYasser) [inline interfaces/builtin/snap_interfaces_requests_control.go:24]: This is a nitpick for a follow up PR, according to https://github.com/canonical/snapd/blob/master/interfaces/builtin/README.md#base-declaration-policy-patterns the plug declaration can be removed entirely.

The same can be said for other similae interfaces `snap-refresh-observe` and `snap-themes-control`.

**Review by ZeyadYasser — APPROVED** [review]
Thank you!

File: interfaces/builtin/snap_interfaces_requests_control.go
Comment (olivercalder) [reply interfaces/builtin/snap_interfaces_requests_control.go:24]: I think this might change when we add the `"handler-service"` attribute in the follow-up PR, so I'll make sure to request you :)

Comment (pedronis) [reply interfaces/builtin/snap_interfaces_requests_control.go:24]: this isn't true because allow-installation is only for the relevant-side

**Pull request merged**
